Use monotonically increasing generation to prevent equivalence cache race

cofyc opened the pull request:
**What this PR does / why we need it**:

Use monotonically increasing generation to prevent equivalence cache race.

Because invalidating predicate cache invalidates latest cache, but in-flight predicate goroutines may write stale results into cache. This may happen when events occur frequently:

| objectVersion | invalidation goroutine | predicate goroutine | cacheVersion | status |
| -- | -- | -- | -- | -- |
| v0 -> v1 | cache invalidated | begin to calculate result.v1 | cache.v0 | OK |
| v1 -> v2 | cache invalidated | begin to write result.v1 into cache | no cache | OK |
| v2 |  | result.v1 wrote into cache | cache.v1 | cache mismatch |

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #67260

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Use monotonically increasing generation to prevent scheduler equivalence cache race.
```

cofyc commented:
@kubernetes/sig-scheduling-pr-reviews 
/assign @msau42

cofyc commented:
/test pull-kubernetes-integration
/test pull-kubernetes-kubemark-e2e-gce-big
/test pull-kubernetes-bazel-test
/test pull-kubernetes-e2e-gce-100-performance

msau42 commented:
/assign @resouer @misterikkit

yastij commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -68,12 +68,16 @@ func NewCache() *Cache {
 	cache predicateMap
+	// Flags which indicate if we received invalidation requests for each
+	// predicate during check was in flight.
+	receivedInvalidatioRequests map[string]bool
```
Rename to `receivedInvalidationRequests`

cofyc replied:
Fixed.

yastij reviewed (commented):
Just a nit

misterikkit commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -282,6 +291,7 @@ func (n *NodeCache) lookupResult(
 	n.mu.RLock()
 	defer n.mu.RUnlock()
 	value, ok = n.cache[predicateKey][equivalenceHash]
+	delete(n.receivedInvalidatioRequests, predicateKey)
```
It seems like anybody who does a lookup will clear the flag for that predicate, causing the next updater to assume that the cache has not been invalidated.

misterikkit commented:
Hi, have you confirmed that this error can happen? Bonus points if you can add a test that triggers the error.

The reason I ask is that we tried to address this specific problem in https://github.com/kubernetes/kubernetes/pull/63040. Perhaps

/cc @bsalamat

cofyc commented:
@misterikkit Add a integration test in e3f6c61d10342cd218f8741ea403bc509c0a98e2.
This test will fail without this feature.

cofyc commented:
/test pull-kubernetes-kubemark-e2e-gce-big
/test pull-kubernetes-e2e-gce-100-performance

cofyc commented:
/test pull-kubernetes-e2e-kops-aws

resouer commented:
/test pull-kubernetes-e2e-kops-aws

resouer commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -282,6 +291,7 @@ func (n *NodeCache) lookupResult(
 	n.mu.RLock()
 	defer n.mu.RUnlock()
 	value, ok = n.cache[predicateKey][equivalenceHash]
+	delete(n.receivedInvalidationRequests, predicateKey)
```
`lookupResult` is in the critical path during scheduling, but this delete operation requires `mu.Lock()` instead of `mu.RLock()`.
 
We may need evaluation of performance impact here.

cofyc replied:
With new solution misterikkit suggested, `RW` lock is not required now.

misterikkit commented:
Just to add some more context, I believe you when you say this is a problem, but I don't believe this PR is the right fix. We tried to fix this exact problem in https://github.com/kubernetes/kubernetes/pull/63040, so my assumption is that something about that fix needs to be corrected.

misterikkit reviewed (commented):
As I mentioned in other comments, I would like this fix to replace the fix for https://github.com/kubernetes/kubernetes/issues/62921

In particular, I think we should,
- add atomic generation numbers to the eCache
- snapshot the current generation numbers before snapshotting the scheduler cache
- compare generation number in snapshot to the live version before writing to eCache
- "undo" the fix from https://github.com/kubernetes/kubernetes/pull/63040 and https://github.com/kubernetes/kubernetes/pull/63895

Sorry that is a lot more work than your initial approach. Let me know if you need help with it. I am also available to chat if you want to discuss the fix.

cofyc commented:
Thanks for clarification and the advice! I'll update this PR.

bsalamat commented:
Thinking more about the solution that misterikkit outlined above last night, I am a bit worried that snapshotting the ecache might cause a noticeable slow down. We should measure performance impact of this solution.

yastij commented:
/retest

cofyc commented:
@misterikkit @bsalamat 

hi, I've updated PR to use solution that misterikkit outlined above. It's ready to review.

Because we don't have any mechanism to notify in-flight goroutine to exit on invalidation, the most important thing to prevent it from writing stale data into cache is we need some way to check the state is changed or not on write. There are two kinds of invalidation:

- invalidate predicate per node (`InvalidatePredicates`, `InvalidatePredicatesOnNode`)
- invalidate node (`InvalidateAllPredicatesOnNode`)

For invalidating predicate per node, add a generation for each predicate on node, increment it by 1 on invalidation, then we can check if there is invalidation happened for this predicate since last looked up.

For invalidating node, deleting node cache is enough. It is possible that in-flight predicate checker may write stale data into node cache, but node cache itself is deleted from equivalence cache. Even if node is added again shortly, scheduler will use fresh node cache next run.

I am not sure how to measure performance impact. Compared to `receivedInvalidationRequests`, it does not require RW lock on `lookupResult`, performance will not be significantly impacted in my opinion. 

Summary of changes:

- revert #63895
- revert part of #63040, `TestCacheInvalidationRace` kept and updated
  - my [integration test](https://github.com/cofyc/kubernetes/commit/d4b85b3001f4b4e603ac5182490fc71f52f6d147) is removed, it does almost the same thing as `TestCacheInvalidationRace`
- update `lookupResult` and `updateResult` to use generation

cofyc commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -258,6 +262,11 @@ func (n *NodeCache) updateResult(
 
 	n.mu.Lock()
 	defer n.mu.Unlock()
+	if received, ok := n.receivedInvalidationRequests[predicateKey]; ok && received {
```
`if !cache.IsUpToDate(nodeInfo)` is removed now, but `nodeInfo` argument is kept, because it's used in other places.

cofyc commented:
/test pull-kubernetes-kubemark-e2e-gce-big

misterikkit commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -300,6 +347,11 @@ func (n *NodeCache) invalidatePreds(predicateKeys sets.String) {
 	defer n.mu.Unlock()
 	for predicateKey := range predicateKeys {
 		delete(n.cache, predicateKey)
+		if _, ok := n.generations[predicateKey]; ok {
```
Similar to above, you can rely on the default map behavior here.
e.g.

```go
m := make(map[string]int)
fmt.Println(m) //  "map[]"
m["counter"]++
fmt.Println(m) // "map[counter:1]"
```

cofyc replied:
Fixed.

misterikkit commented on pkg/scheduler/core/equivalence/eqivalence_test.go:
```diff
@@ -190,31 +180,23 @@ func TestRunPredicate(t *testing.T) {
 			expectCacheWrite: true,
 		},
-		{
-			name:             "pod fits/cache miss/no write",
```
Is there a new test that covers this scenario?

cofyc replied:
This test is kept and updated now.
Found that `schedulercache.Cache` can be replaced by `generation` field.

misterikkit commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -95,6 +95,7 @@ func (f *FitError) Error() string {
 type genericScheduler struct {
 	cache                    schedulercache.Cache
 	equivalenceCache         *equivalence.Cache
+	nodeCacheSnapshots       map[string]*equivalence.NodeCacheSnapshot
```
I think this member should be named so it's clearly related to eCache. However, I am not convinced that we should have this.

The reason that we have `cachedNodeInfoMap` is so that we can avoid rebuilding a snapshot from scratch every time. In `UpdateNodeNameToInfoMap`, we check the generation ID and decide whether we need to DeepCopy the NodeInfo object. Since eCache snapshots are just generation IDs, the only thing you (potentially) avoid is map allocation. (might be worth benchmarking with and without...)

cofyc replied:
I added this only to avoid map allocation. I'll do some benchmarks. By the way I'm trying to figure out is it possible to use sequence numbers to present predicate key, then we can use slices to store generations ids and predicate results also. In that case, performance will be much better.

bsalamat replied:
I think if the snapshot was a member of equivalenceCache, you didn't need to pass it separately here.

misterikkit commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -129,6 +130,11 @@ func (g *genericScheduler) Schedule(pod *v1.Pod, nodeLister algorithm.NodeLister
 		return "", ErrNoNodesAvailable
 	}
 
+	// Snapshot all predicate generations and references of node caches.
```
It is important that this happens before we snapshot the regular cache. (`UpdateNodeNameToInfoMap` is not the best name for that...) Could you expand on this comment so that future readers know the order matters? (Maybe borrow wording from the deleted comments in factory.go.)

In case it is not clear, the reason we must have it this way is the following scenario.
1. snapshot cache
2. event arrives, updating cache and invalidating ecache
3. snapshot ecache
4. evaluate predicates
5. result is written to ecache... using stale info from the cache snapshot

cofyc replied:
Updated but not sure if it's good enough. Let me know if you have some suggestions about wording.

misterikkit commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -564,7 +576,11 @@ func podFitsOnNode(
 			if predicate, exist := predicateFuncs[predicateKey]; exist {
 				if eCacheAvailable {
-					fit, reasons, err = nodeCache.RunPredicate(predicate, predicateKey, pod, metaToUse, nodeInfoToUse, equivClass, cache)
+					predicateGeneration := uint64(0)
```
Since zero values are default, the preferred (canonical) way to write this is `var predicateGeneration uint64`.

misterikkit commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -564,7 +576,11 @@ func podFitsOnNode(
 				if eCacheAvailable {
-					fit, reasons, err = nodeCache.RunPredicate(predicate, predicateKey, pod, metaToUse, nodeInfoToUse, equivClass, cache)
+					predicateGeneration := uint64(0)
+					if predicateGenerations != nil {
```
Actually, it's perfectly valid to do a lookup on a nil map. You will get the zero value.

cofyc replied:
Updated, and put it into `nodeCache.RunPredicate(...)`.

misterikkit commented on pkg/scheduler/core/generic_scheduler_test.go:
```diff
@@ -1504,3 +1504,90 @@ func TestCacheInvalidationRace(t *testing.T) {
 }
+
+// TestCacheInvalidationRace2 tests that cache invalidation is correctly handled
+// when an invalidation event happens during predicate is running.
```
grammar nit: "during predicate is running" should be "while a predicate is running"

cofyc replied:
Updated.

misterikkit commented on pkg/scheduler/core/generic_scheduler_test.go:
```diff
@@ -1504,3 +1504,90 @@ func TestCacheInvalidationRace(t *testing.T) {
+// when an invalidation event happens during predicate is running.
+func TestCacheInvalidationRace2(t *testing.T) {
+	// Create a predicate that returns false the first time and true on subsequent calls.
+	podWillFit := false
```
should this be in the var section?

cofyc replied:
Yes, updated.

misterikkit commented on pkg/scheduler/core/generic_scheduler_test.go:
```diff
@@ -1504,3 +1504,90 @@ func TestCacheInvalidationRace(t *testing.T) {
+			failureReasons []algorithm.PredicateFailureReason
+		)
+		if !podWillFit {
+			podWillFit = true
```
You could shorten this function by moving once.Do to the top and returning inside the conditional statement.

cofyc replied:
Updated.

misterikkit reviewed (commented):
Looking pretty good. I made an initial pass over the code, but I haven't given it a thorough review yet.

misterikkit commented:
To run the benchmarks, I suggest reading through https://github.com/kubernetes/kubernetes/pull/65714#issuecomment-410016382 to see the commands that @bsalamat  and @resouer were using, and tweak that as needed.

cofyc commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -206,10 +246,11 @@ type predicateResult struct {
 //
-// NOTE: RunPredicate will not update the equivalence cache if the given NodeInfo is stale.
+// NOTE: RunPredicate will not update the equivalence cache if generation does not match live version.
 func (n *NodeCache) RunPredicate(
```
Updated.

cofyc commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -262,6 +299,16 @@ func (n *NodeCache) updateResult(
 	defer n.mu.Unlock()
+	liveGeneration, ok := n.generations[predicateKey]
+	if !ok {
+		liveGeneration = 0
```
Updated. Didn't know this, thanks!

cofyc commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -262,6 +299,16 @@ func (n *NodeCache) updateResult(
+		// Generation of this predicate has been updated since we last looked
+		// up, this indicates that we received a invalidation request during
+		// this time. Cache may be stale, skip update.
+		return
```
Fixed. My merge mistake too.

resouer commented:
I've read thru this PR and it generally looks good. While my concern is also same with Bobby: what's the performance impact?

> maybe we can order predicates and use sequence numbers to represent predicate key internally?

AFAIK, predicates are already ordered, see: https://github.com/kubernetes/kubernetes/blob/5a16163c87fe2a90916a51b52771a668bcaf2a0d/pkg/scheduler/core/generic_scheduler.go#L558 

> With this, we can use slice (or array) to store predicate results and generations in equivalence cache. Accessing and copying items in slice have much better performance than map.

I will +1 for this idea. 

It's not a secret that Golang map access used in ecache is a issue, and as the result, ecache becomes slower when the predicate itself is pure mathematics operation. I am very interested for cheap predicates, when ecache=ture, benchmark results should be at least the same with ecache=false Could you please benchmark this as well? (No need to send out PR, just to see if it's worthy)

P.S. Please use the benchmarking command below with parameter change FYI:
```
make test-integration WHAT=./test/integration/scheduler_perf GOFLAGS="-v=1" KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-bench=BenchmarkScheduling -cpuprofile=cpu-5000n-10000p.out"
```
```
diff --git a/test/integration/scheduler_perf/scheduler_bench_test.go b/test/integration/scheduler_perf/scheduler_bench_test.go
index fb28da4..f3bc6d8 100644
--- a/test/integration/scheduler_perf/scheduler_bench_test.go
+++ b/test/integration/scheduler_perf/scheduler_bench_test.go
@@ -35,10 +35,7 @@ import (
 // various quantities of nodes and scheduled pods.
 func BenchmarkScheduling(b *testing.B) {
 	tests := []struct{ nodes, existingPods, minPods int }{
-		{nodes: 100, existingPods: 0, minPods: 100},
-		{nodes: 100, existingPods: 1000, minPods: 100},
-		{nodes: 1000, existingPods: 0, minPods: 100},
-		{nodes: 1000, existingPods: 1000, minPods: 100},
+		{nodes: 5000, existingPods: 0, minPods: 10000},
 	}
 	setupStrategy := testutils.NewSimpleWithControllerCreatePodStrategy("rc1")
 	testStrategy := testutils.NewSimpleWithControllerCreatePodStrategy("rc2")
diff --git a/test/integration/util/util.go b/test/integration/util/util.go
index 705abbf..6b43520 100644
--- a/test/integration/util/util.go
+++ b/test/integration/util/util.go
@@ -95,7 +95,7 @@ func createSchedulerConfigurator(
 	informerFactory informers.SharedInformerFactory,
 ) scheduler.Configurator {
 	// Enable EnableEquivalenceClassCache for all integration tests.
-	utilfeature.DefaultFeatureGate.Set("EnableEquivalenceClassCache=true")
+	utilfeature.DefaultFeatureGate.Set("EnableEquivalenceClassCache=false")
 
 	return factory.NewConfigFactory(
 		v1.DefaultSchedulerName,
-- 
```

resouer commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -70,15 +70,55 @@ func NewCache() *Cache {
+// UpdateNodeCacheSnapshots update node cache snapshots for given nodes. Create node cache if does not exist.
+func (c *Cache) UpdateNodeCacheSnapshots(nodes []*v1.Node, snapshots map[string]*NodeCacheSnapshot) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
```
In recent #66862, we've changed to use `sync.Map` for `Cache`, could you please rebase to latest HEAD and use this data structure?

cofyc replied:
Found it's impossible to wrap more than 1 operations on map in a lock, e.g. updating and iterating.

In this snapshotting function, we need to create node cache if not found, then snapshot all node caches. I updated it to use `sync.Map` methods (see below), apparently it's not ok. 

- cannot wrap multiple update operations in one lock
- cannot wrap update operations with iterations in one lock

```
    for _, node := range nodes {
        if _, ok := c.Load(node.Name); !ok {
            c.LoadOrStore(node.Name, newNodeCache())
        }
    }
    c.Range(func(k, v interface{}) bool {
        name := k.(string)
        node := v.(*NodeCache)
        snapshot, ok := snapshots[name]
        if !ok {
            snapshot = &NodeCacheSnapshot{
                Generations: make(map[string]uint64, len(node.generations)),
            }
            snapshots[name] = snapshot
        }
        snapshot.NodeCache = node
        for k, v := range node.generations {
            snapshot.Generations[k] = v
        }
        return true
    })
```

Still using plain Go map with separate locking, what do you think? I think it gives us more freedom.

resouer replied:
I'm totally OK with either way. The only concern will be performance decrease in machines with 8+ CPUs, let's see how the benchmark says :D

resouer commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -70,15 +70,55 @@ func NewCache() *Cache {
 	cache predicateMap
+	// Generations for predicates, incremented on predicate invalidation.
+	// Created on first update. Use 0 if does not exist.
+	generations map[string]uint64
```
The typical comment style is: "generations is ..."

cofyc replied:
Updated.

misterikkit commented:
> P.S. Please use the benchmarking command below with parameter change FYI:

One of these days we should figure out how to make these parameters so we can run tests without carrying patches around.

Commit:
Revert "Use sync.map to scale ecache better"

This reverts commit 17d0190706407bade8f6f875844749007904ad7d.

bsalamat commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -63,15 +63,57 @@ func NewCache() *Cache {
 
+// NodeCacheSnapshot represents a snapshot of NodeCache reference and generations.
+type NodeCacheSnapshot struct {
+	NodeCache   *NodeCache
```
Why do we need to store this pointer? We just need to snapshot the generation number.

cofyc replied:
I have some explanations [here](https://github.com/kubernetes/kubernetes/pull/67308#issuecomment-414054511). For invalidating node, deleting node cache is not enough, because node is invalidated before `lookupResult`, a fresh node cache will be created in `lookupResult` and stale data will be written into it. To fix this, we can get node cache before snapshotting scheduler cache. If in-flight predicate checker write stale data into it, node cache itself can be removed from equivalence cache on invalidation, so scheduler will use fresh node cache next run.

I prefer this because 1) we don't need to store node cache forever even after node is deleted from cluster 2) checkNode needs to get node cache one by one, getting node cache pointers all at once does not add performance overhead.

cofyc replied:
@misterikkit @resouer What do you think?

bsalamat commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -63,15 +63,57 @@ func NewCache() *Cache {
+	// generations stores generation numbers for predicates, incremented on
+	// predicate invalidation. Created on first update. Use 0 if does not
+	// exist.
+	generations map[string]uint64
```
I think you can add the snapshot as a member of NodeCache.

cofyc replied:
IIUC, snapshotting generations to current node cache, then in `updateResult` compare previous snapshotted generation with live generation?
It's ok, but requires us to make assumptions how functions are called.

cofyc replied:
@misterikkit @resouer What do you think?

bsalamat commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -287,6 +330,7 @@ func (n *NodeCache) invalidatePreds(predicateKeys sets.String) {
 	defer n.mu.Unlock()
 	for predicateKey := range predicateKeys {
 		delete(n.cache, predicateKey)
+		n.generations[predicateKey]++
```
Also, at line 153, all predicates of a node are invalidated. We currently just delete the node from the cache. With this PR, we can no longer delete the node. We must keep the node, increment generation number of all the predicates and set the `predicateMap` to nil (or an empty map).

bsalamat reviewed (commented):
Thanks, @cofyc! I have several early comments.

cofyc commented:
Here is my benchmark results, some are still work in progress. I reverted #66862 for this PR because [the reasons I listed here](https://github.com/kubernetes/kubernetes/pull/67308#discussion_r212554374).

## Benchmarks (WIP)

### Profile

Machine:

- CPUs: 8 x Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz

Benchmark configurations:

- BenchmarkScheduling: `{nodes: 5000, existingPods: 0, minPods: 10000}`
- BenchmarkSchedulingAntiAffinity: `{nodes: 2500, existingPods: 0, minPods: 5000}`

Benchmark command:

```
./hack/print-workspace-status.sh | grep '^gitCommit'
grep -o -P 'EnableEquivalenceClassCache=(true|false)' test/integration/util/util.go
make test-integration WHAT=./test/integration/scheduler_perf GOFLAGS="-v=1" KUBE_TEST_VMODULE="''" KUBE_TIMEOUT="--timeout 3600s" KUBE_TEST_ARGS="-bench=BenchmarkScheduling" | tee benchmark.log | grep -P '^(BenchmarkScheduling|ok)'
```

### N5000-P10000-ECACHE-ENABLED-PR67308

```
gitCommit 6b59be11821abf6c630f55c4759f43dc3afa83a8
EnableEquivalenceClassCache=true
BenchmarkScheduling/5000Nodes/0Pods-8         	   10000	  18826094 ns/op
BenchmarkSchedulingAntiAffinity/2500Nodes/0Pods-8         	    5000	  16425119 ns/op
ok  	k8s.io/kubernetes/test/integration/scheduler_perf	285.792s
```

### N5000-P10000-ECACHE-DISABLED-PR67308

```
gitCommit 6b59be11821abf6c630f55c4759f43dc3afa83a8
EnableEquivalenceClassCache=false
BenchmarkScheduling/5000Nodes/0Pods-8         	   10000	  11951690 ns/op
BenchmarkSchedulingAntiAffinity/2500Nodes/0Pods-8         	    5000	 142623149 ns/op
ok  	k8s.io/kubernetes/test/integration/scheduler_perf	848.066s
```

### N5000-P10000-ECACHE-ENABLED-MASTER

```
gitCommit 5b2a7c7347cf9ca86ef5dfaaea3e3aa4df3ae270
EnableEquivalenceClassCache=true
BenchmarkScheduling/5000Nodes/0Pods-8         	   10000	  15743979 ns/op
BenchmarkSchedulingAntiAffinity/2500Nodes/0Pods-8         	    5000	  14195174 ns/op
ok  	k8s.io/kubernetes/test/integration/scheduler_perf	243.524s
```

### N5000-P10000-ECACHE-DISABLED-MASTER

```
gitCommit 5b2a7c7347cf9ca86ef5dfaaea3e3aa4df3ae270
EnableEquivalenceClassCache=false
BenchmarkScheduling/5000Nodes/0Pods-8         	   10000	  11452931 ns/op
BenchmarkSchedulingAntiAffinity/2500Nodes/0Pods-8         	    5000	 130701043 ns/op
ok  	k8s.io/kubernetes/test/integration/scheduler_perf	659.675s
```

### N5000-P10000-ECACHE-ENABLED-PR67308-INTKEY (TODO)

### N5000-P10000-ECACHE-DISABLED-PR67308-INTKEY (TODO)

cofyc commented:
/test pull-kubernetes-integration

misterikkit reviewed (changes requested):
Oops, I wrote these comments last week and forgot to hit send.

cofyc commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -70,15 +70,55 @@ func NewCache() *Cache {
 	}
 }
 
+// NodeCacheSnapshot represents a snapshot of NodeCache reference and generations.
```
Updated.

cofyc commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -262,6 +295,14 @@ func (n *NodeCache) updateResult(
 	defer n.mu.Unlock()
+	liveGeneration := n.generations[predicateKey]
+	if generation != liveGeneration {
+		// Generation of this predicate has been updated since we last looked
```
oops, this comment is for my initial implementation, in which I cached generations in `lookupResult`. Fixed now.

cofyc commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -277,8 +318,8 @@ func (n *NodeCache) updateResult(
 
-// lookupResult returns cached predicate results and a bool saying whether a
-// cache entry was found.
+// lookupResult returns cached predicate results with a associated generation
```
Fixed.

cofyc commented:
@misterikkit @bsalamat @resouer 

In my benchmark results, I've tried some alternative implementations:

### Use int key to present predicate key

resouer commented:
Thanks for so much detailed benchmark.

One nits I need to amend is (sorry, just noticed it recently) the metrics calculation in `lookupResult` is not cheap, (see: #67857), so the result of ecache enabled is actually inaccurate, but it should not change the result as long as all the tests included the metrics patch. 

Personally, I am in favor of a seperate PR to use Integer as key of ecache, and would like to do a review :) (You can revert sync.Map there too)

> Another change in this implementation is to invalidate node cache instead of caching its pointer, but this requires to keep node cache even after node is removed from cluster

This does not seems to be a big problem to me.

 btw, using Integer as predicate key does not conflict with Bobby's suggestion, right?

cofyc commented:
> btw, using Integer as predicate key does not conflict with Bobby's suggestion, right?

Yes, using Integer as predicate key is independent of all choices.

cofyc commented:
hi, I pushed a new commit to address @bsalamat's suggestions. PTAL.

cofyc commented:
/test pull-kubernetes-e2e-kops-aws

cofyc commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -389,7 +401,9 @@ func (g *genericScheduler) findNodesThatFit(pod *v1.Pod, nodes []*v1.Node) ([]*v
 		checkNode := func(i int) {
-			var nodeCache *equivalence.NodeCache
+			var (
+				nodeCache *equivalence.NodeCache
```
this place has been edited multiple times, no change now, I'll revert it to original state

cofyc commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -61,23 +68,56 @@ func NewCache() *Cache {
+	defer c.mu.Unlock()
+	for _, node := range nodes {
+		if _, exists := c.nodeToCache[node.Name]; !exists {
+			c.nodeToCache[node.Name] = newNodeCache()
```
I thought earlier we must make sure all nodes has node caches because equivalence cache may not have caches for [nodes listed from lister](https://github.com/kubernetes/kubernetes/blob/v1.13.0-alpha.0/pkg/scheduler/core/generic_scheduler.go#L115) and `checkNode` will get a fresh node cache to use. Invalidations happens after snapshotting and before node cache being created will be ignored. (This is what `TestCacheInvalidationRace` tests)

But I find now it's not necessary because we run predicates on nodes in scheduler cache not nodes listed from lister after [this PR](https://github.com/kubernetes/kubernetes/pull/66733/files#diff-c237cdd9e4cb201118ca380732d7f361L367). When a node is added into scheduler cache, [it will create node cache for it](https://github.com/kubernetes/kubernetes/blob/e124159990c6d22e9ca662ef35de5eeaef5d8956/pkg/scheduler/factory/factory.go#L856).

I'll update this and move `GetNodeCache()` before `schedulerCache.AddNode` in `configFactory. addNodeToCache()` to make sure node cache is created before being adding into schedule cache.

resouer replied:
> I'll update this and move GetNodeCache() before schedulerCache.AddNode in configFactory. addNodeToCache() to make sure node cache is created before being adding into schedule cache.

This is what I'm thinking recently: we should try to avoid calling`GetNodeCache()` in any scheduling path as it holds a lock across nodes. You need to handle in node delete as well.

cofyc replied:
@resouer Thanks! We should invalidate node before removing it from scheduler cache.

cofyc commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -61,23 +68,56 @@ func NewCache() *Cache {
 func (c *Cache) GetNodeCache(name string) (nodeCache *NodeCache, exists bool) {
-	v, exists := c.LoadOrStore(name, newNodeCache())
-	nodeCache = v.(*NodeCache)
+	c.mu.Lock()
```
In earlier implementation, I need to create and snapshot node cache in one function, but `sync.Map` does not support more than 1 operations in one lock. I'll revert it back.

cofyc commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -249,6 +284,13 @@ func (n *NodeCache) updateResult(
+		return
+	}
 	// If cached predicate map already exists, just update the predicate by key
 	if predicates, ok := n.cache[predicateKey]; ok {
```
oops, my mistake, updated

cofyc commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -129,6 +129,18 @@ func (g *genericScheduler) Schedule(pod *v1.Pod, nodeLister algorithm.NodeLister
+	// 4. evaludate predicates
+	// 5. stale result will be written to ecache
+	if g.equivalenceCache != nil {
+		g.equivalenceCache.Snapshot(nodes)
```
Updated.

cofyc commented:
/test pull-kubernetes-integration
/test pull-kubernetes-e2e-gce

bsalamat commented:
@cofyc Thanks a lot for running the benchmarks and collecting information for various implementation options. The performance improvement from the predicate int keys is great. I think we should use it instead of the current implementation.

cofyc commented:
/test pull-kubernetes-kubemark-e2e-gce-big

cofyc commented:
Any other concerns?

bsalamat commented on pkg/scheduler/factory/factory.go:
```diff
@@ -849,15 +843,17 @@ func (c *configFactory) addNodeToCache(obj interface{}) {
-		glog.Errorf("scheduler cache AddNode failed: %v", err)
-	}
-
+	// NOTE: Because the scheduler uses equivalence caches for nodes, we need
```
s/caches/cache/

cofyc replied:
fixed

bsalamat commented on pkg/scheduler/factory/factory.go:
```diff
@@ -970,15 +963,13 @@ func (c *configFactory) deleteNodeFromCache(obj interface{}) {
 	if c.enableEquivalenceClassCache {
 		c.equivalencePodCache.InvalidateAllPredicatesOnNode(node.GetName())
 	}
+	if err := c.schedulerCache.RemoveNode(node); err != nil {
```
If remove must happen after the invalidation, it deserves a better comment.

cofyc replied:
Sorry, I found I'm wrong with this part. All equivalence cache invalidations should be after resource objects are added, updated and deleted, then predicate functions will run and use new versions of resource objects. I revert this change.

By the way, there is no need to invalidate predicates for node after adding because no predicate functions will run on deleted node (we iterate on nodes in scheduler cache).

bsalamat replied:
I agree with you that invalidation of eCache should happen after CUD operations on objects. The following diagram shows a scenario where the reverse order causes an issue. If invalidation happens first, we could snapshot eCache after the invalidation and then we snapshot the cache itself. A node is then removed/added/updated before we update eCache. We then would update eCache with stale information which is based on our latest snapshot of the cache taken before the node removal/addition/update:

```
------|-----------------|------------------|----------------|---------------|-------->
Invalidate eCache  Snapshot eCache   Snapshot Cache     Remove Node    Update eCache    
```

I think you should add this to the comment to document why the order is important.

cofyc replied:
Added in https://github.com/kubernetes/kubernetes/pull/67308/commits/dd9d576e64f4536ffe8d9367d1d2edc90ee3d3ab.

There is no need to revert PR https://github.com/kubernetes/kubernetes/pull/63895/files, I'll rebase it when PR is done.

bsalamat reviewed (commented):
Thanks, @cofyc and sorry for the delay. It looks generally good. I have several minor comments.
I haven't looked at the tests yet.

cofyc commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -61,36 +74,78 @@ func NewCache() *Cache {
+	for predicateKey := range predicateKeys {
+		if id, ok := c.predicateIDMap[predicateKey]; ok {
+			predicateIDs = append(predicateIDs, id)
+		}
```
added

cofyc commented:
@bsalamat Thanks, I pushed a new commit: https://github.com/kubernetes/kubernetes/pull/67308/commits/cc3b6d9da3e0d77b9e9376cbf39c5b1e7b7518b4.

resouer commented on pkg/scheduler/factory/factory.go:
```diff
@@ -849,15 +843,17 @@ func (c *configFactory) addNodeToCache(obj interface{}) {
+	// to create it before adding node into scheduler cache.
 	if c.enableEquivalenceClassCache {
 		// GetNodeCache() will lazily create NodeCache for given node if it does not exist.
 		c.equivalencePodCache.GetNodeCache(node.GetName())
```
Since we will init NodeCache here, it seems we don't need to call `g.equivalenceCache.GetNodeCache(nodeName)` in generic_scheduler?

Instead, we can only call sth like `g.equivalenceCache.LoadNodeCache()` there, only read operation. WDYT?

resouer commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -35,20 +35,33 @@ import (
 	"github.com/golang/glog"
 )
 
+// nodeMap stores a *Cache for each node.
```
stores a *NodeCache

resouer commented on pkg/scheduler/core/equivalence/eqivalence.go:
```diff
@@ -61,36 +74,80 @@ func NewCache() *Cache {
+	// predicate invalidation. Created on first update. Use 0 if does not
+	// exist.
+	predicateGenerations         []uint64
+	snapshotPredicateGenerations []uint64
```
It would be better to include comments for all last three members, to make other developer's life easier.

resouer reviewed (commented):
Just added several review comments, others look good, please take a look!

bsalamat commented on pkg/scheduler/scheduler.go:
```diff
@@ -329,9 +329,11 @@ func (sched *Scheduler) assume(assumed *v1.Pod, host string) error {
-	// NOTE: Because the scheduler uses snapshots of SchedulerCache and the live
-	// version of Ecache, updates must be written to SchedulerCache before
-	// invalidating Ecache.
+	// NOTE: Updates must be written to scheduler cache before invalidating
```
I don't think this comment applies here.

cofyc replied:
After assuming pod into scheduler cache, it will invalidate affected predicates in equivalence cache.
https://github.com/kubernetes/kubernetes/blob/8ece5d624ded4815655104d8a12712cfe6d8743e/pkg/scheduler/scheduler.go#L355-L363

bsalamat reviewed (approved):
Thanks, @cofyc!

/lgtm

Commit:
Update notes to document why invalidation order is important.

Commit:
Use seqeuence number to represent generation of equivalence cache.

- snapshot equivalence cache generation numbers before snapshotting the
scheduler cache
- skip update when generation does not match live generation
- keep the node and increment its generation to invalidate it instead of
deletion
- use predicates order ID as key to improve performance

cofyc commented:
/retest

cofyc commented:
/retest

cofyc commented:
/retest

cofyc commented:
/retest

cofyc commented:
/retest

bsalamat reviewed (commented):
/lgtm
